### PR TITLE
Adding JenkinsEventNotifier to send events back to TFS/VSTS

### PR DESCRIPTION
**Pull request by jpricketMSFT**
Adding "team-events-connect" endpoint which allows TFS/VSTS to "connect" to the Jenkins server and start receiving events.

### CODE REVIEW

File: tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+import java.util.List;
+import java.util.logging.Logger;
+
+/**
+ * This class sends Jenkins events to all "connected" team collections.
+ * Current list of events:
+ * - Job Completion event
+ */
+public class JenkinsEventNotifier {
+    protected final static Logger log = Logger.getLogger(JenkinsEventNotifier.class.getName());
```
Comment (lkillgore) [inline tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java]: (I don't know if we care): I think Checkstyle will flag this line for the order of the keywords:

https://stackoverflow.com/questions/10299067/modifier-keyword-order-in-java

File: tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+                final ConnectionParameters connectionParameters = c.getConnectionParameters();
+                if (connectionParameters == null || !connectionParameters.isSendJobCompletionEvents()) {
+                    // This server isn't accepting events
+                    continue;
+                }
+                final TeamRestClient client = new TeamRestClient(URI.create(c.getCollectionUrl()));
+                payload.put("server", connectionParameters.getConnectionKey());
+                final String jsonPayload = payload.toString();
+                final JobCompletionEventArgs args = new JobCompletionEventArgs(
+                        connectionParameters.getConnectionKey(), jsonPayload,
```
Comment (AlexRukhlin) [inline tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java]: Split this line ?

File: tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static String urlCombine(final String url, final String... parts) {
+        final StringBuilder sb = new StringBuilder();
+        sb.append(url);
+        for(final String s : parts) {
+            if (StringUtils.isNotBlank(s)) {
+                if (sb.charAt(sb.length() - 1) != '/') {
```
Comment (lkillgore) [inline tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java]: Protect against an empty url?

File: tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java (hunk 1 above)
Comment (jpricketMSFT) [inline tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java]: Probably right. I will run check style and fix those issues.

File: tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java (hunk 2 above)
Comment (jpricketMSFT) [inline tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java]: I assume you want me to put each parameter on a separate line? That makes sense.

File: tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java (hunk 3 above)
Comment (jpricketMSFT) [inline tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java]: good catch.

File: tfs/src/main/java/hudson/plugins/tfs/listeners/JenkinsRunListener.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.logging.Logger;
+
+/**
+ * This class listens to the events of every Jenkins run instance.
+ * Completed runs fire an event back to the JenkinsEventNotifier.
+ */
+@Extension
+public class JenkinsRunListener extends RunListener<Run> {
+
+    protected static Logger log = Logger.getLogger(JenkinsRunListener.class.getName());
```
Comment (lkillgore) [inline tfs/src/main/java/hudson/plugins/tfs/listeners/JenkinsRunListener.java]: +final

File: tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java (hunk 3 above)
Comment (AlexRukhlin) [inline tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java]: Shouldn't we check that sb.Length() > 0 ?

File: tfs/src/main/java/hudson/plugins/tfs/listeners/JenkinsRunListener.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        final JSONObject json = JSONObject.fromObject(payload);
+        json.put("name", run.getParent().getDisplayName());
+        json.put("startedBy", getStartedBy(run));
+
+        JenkinsEventNotifier.sendJobCompletionEvent(json);
+    }
+
+    private String getStartedBy(final Run run) {
+        final Cause.UserIdCause cause = (Cause.UserIdCause) run.getCause(Cause.UserIdCause.class);
+        final String startedBy;
```
Comment (lkillgore) [inline tfs/src/main/java/hudson/plugins/tfs/listeners/JenkinsRunListener.java]: I think this would be simpler without the 'final' and the 'else' clause.  Just initialize to 'null'.

File: tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    /**
+     * Send the Job Completion event to connected TFS/VSTS servers
+     */
+    public static void sendJobCompletionEvent(final JSONObject payload) {
+        final List<TeamCollectionConfiguration> connectedCollections = TeamCollectionConfiguration.getConnectedCollections();
+        for (final TeamCollectionConfiguration c : connectedCollections) {
+            try {
+                // Check to see if there are any collections "connected" to this Jenkins server
+                final ConnectionParameters connectionParameters = c.getConnectionParameters();
+                if (connectionParameters == null || !connectionParameters.isSendJobCompletionEvents()) {
```
Comment (AlexRukhlin) [inline tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java]: It seems to me that the method TeamCollectionConfiguration.getConnectedCollections() already does this check.

Comment (jpricketMSFT) [inline tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java]: good point. I will remove the null check and perhaps annotate the method with NotNull

File: tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java (hunk 3 above)
Comment (jpricketMSFT) [inline tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java]: absolutely, good catch.

File: tfs/src/main/java/hudson/plugins/tfs/listeners/JenkinsRunListener.java (hunk 4 above)
Comment (jpricketMSFT) [inline tfs/src/main/java/hudson/plugins/tfs/listeners/JenkinsRunListener.java]: yep

File: tfs/src/main/java/hudson/plugins/tfs/TeamEventsEndpoint.java [hunk 7]
```diff
@@ -158,8 +160,8 @@ static JSONObject innerDispatch(final String body, final String eventName, final
         }
         final AbstractHookEvent.Factory factory = factoriesByName.get(eventName);
         final Event serviceHookEvent = deserializeEvent(body);
-        final String message = serviceHookEvent.getMessage().getText();
-        final String detailedMessage = serviceHookEvent.getDetailedMessage().getText();
+        final String message = serviceHookEvent.getMessage() != null ? serviceHookEvent.getMessage().getText() : "";
+        final String detailedMessage = serviceHookEvent.getDetailedMessage() != null ? serviceHookEvent.getDetailedMessage().getText() : "";
```
Comment (AlexRukhlin) [inline tfs/src/main/java/hudson/plugins/tfs/TeamEventsEndpoint.java]: Why do we need these messages?

File: tfs/src/main/java/hudson/plugins/tfs/listeners/JenkinsRunListener.java (hunk 5 above)
Comment (jpricketMSFT) [inline tfs/src/main/java/hudson/plugins/tfs/listeners/JenkinsRunListener.java]: It would be less lines, but I like the final. I will see how it looks.

File: tfs/src/main/java/hudson/plugins/tfs/util/TeamRestClient.java [hunk 8]
```diff
@@ -287,4 +305,27 @@ public TeamGitStatus addPullRequestIterationStatus(final PullRequestMergeCommitC
 
         return request(TeamGitStatus.class, HttpMethod.POST, requestUri, status);
     }
+
+
+    public void sendJobCompletionEvent(final JobCompletionEventArgs args) throws IOException {
+        final QueryString qs = new QueryString(
+                API_VERSION, "3.0-preview.1",
```
Comment (AlexRukhlin) [inline tfs/src/main/java/hudson/plugins/tfs/util/TeamRestClient.java]: 4.0 ?

File: tfs/src/main/java/hudson/plugins/tfs/TeamEventsEndpoint.java (hunk 7 above)
Comment (AlexRukhlin) [inline tfs/src/main/java/hudson/plugins/tfs/TeamEventsEndpoint.java]: It looks like we do not use these messages.

File: tfs/src/main/java/hudson/plugins/tfs/util/TeamRestClient.java (hunk 8 above)
Comment (AlexRukhlin) [inline tfs/src/main/java/hudson/plugins/tfs/util/TeamRestClient.java]: Shouldn't we use a released API version?

File: tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+                }
+                final TeamRestClient client = new TeamRestClient(URI.create(c.getCollectionUrl()));
+                payload.put("server", connectionParameters.getConnectionKey());
+                final String jsonPayload = payload.toString();
+                final JobCompletionEventArgs args = new JobCompletionEventArgs(
+                        connectionParameters.getConnectionKey(), jsonPayload,
+                        getPayloadSignature(connectionParameters, jsonPayload));
+                client.sendJobCompletionEvent(args);
+            } catch (final Exception e) {
+                log.info("ERROR: sendJobCompletionEvent: (collection=" + c.getCollectionUrl() + ") " + e.getMessage());
```
Comment (yacaovsnc) [inline tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java]: Maybe this log level should be "warn"?  Is this error expected?

File: tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java (hunk 1 above)
Comment (jpricketMSFT) [inline tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java]: I added CheckStyle to the build but it will be in another PR since there are so many changes.

File: tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java (hunk 9 above)
Comment (jpricketMSFT) [inline tfs/src/main/java/hudson/plugins/tfs/JenkinsEventNotifier.java]: Agreed. It is not expected. I looked for a log.error, but there wasn't one :(

File: tfs/src/main/java/hudson/plugins/tfs/TeamEventsEndpoint.java (hunk 7 above)
Comment (jpricketMSFT) [inline tfs/src/main/java/hudson/plugins/tfs/TeamEventsEndpoint.java]: I don't see them being used either, but it would be bigger change to remove them. I only changed the code because I ran into an error here. I don't know why they were originally added.

File: tfs/src/main/java/hudson/plugins/tfs/listeners/JenkinsRunListener.java (hunk 4 above)
Comment (jpricketMSFT) [inline tfs/src/main/java/hudson/plugins/tfs/listeners/JenkinsRunListener.java]: I made this change in another PR (with CheckStyle)

File: tfs/src/main/java/hudson/plugins/tfs/util/TeamRestClient.java (hunk 8 above)
Comment (jpricketMSFT) [inline tfs/src/main/java/hudson/plugins/tfs/util/TeamRestClient.java]: will do. Thanks for providing the version information.
